internal/nix/command.go

```diff
@@ -1,20 +1,258 @@
+		c.execCmd.Env = c.Env
+		c.execCmd.Stdin = c.Stdin
+		c.execCmd.Stdout = c.Stdout
+		c.execCmd.Stderr = c.Stderr
+
+		c.execCmd.Cancel = func() error {
+			// Try to let Nix exit gracefully by sending an
+			// interrupt instead of the default behavior of killing
+			// it.
+			err := c.execCmd.Process.Signal(os.Interrupt)
+			if errors.Is(err, os.ErrProcessDone) {
+				// Nix already exited; execCmd.Wait will use the
+				// exit code.
+				return err
+			}
+			if err != nil {
+				// We failed to send SIGINT, so kill the process
+				// instead.
+				//
+				// - If Nix already exited, Kill will return
+				//   os.ErrProcessDone and execCmd.Wait will use
+				//   the exit code.
+				// - Otherwise, execCmd.Wait will always return
+				//   an error.
+				return c.execCmd.Process.Kill()
+			}
+
+			// We sent the SIGINT successfully. It's still possible
+			// for Nix to exit successfully, so return
+			// os.ErrProcessDone so that execCmd.Wait uses the exit
+			// code instead of ctx.Err.
+			return os.ErrProcessDone
+		}
+		// Kill Nix if it doesn't exit within 15 seconds of Devbox
+		// sending an interrupt.
+		c.execCmd.WaitDelay = 15 * time.Second
+	})
+	return c.execCmd
+}
+
+func (c *cmd) error(ctx context.Context, err error) error {
+	if err == nil {
+		return nil
+	}
+
+	cmdErr := &cmdError{err: err}
+	if errors.Is(err, exec.ErrNotFound) {
+		cmdErr.msg = fmt.Sprintf("nix: %s not found in $PATH", c.Args[0])
+	}
+
+	switch {
+	case errors.Is(ctx.Err(), context.Canceled):
+		cmdErr.msg = "nix: command canceled"
+	case errors.Is(ctx.Err(), context.DeadlineExceeded):
+		cmdErr.msg = "nix: command timed out"
+	default:
+		cmdErr.msg = "nix: command error"
+	}
+	cmdErr.msg += ": " + c.String()
+
+	var exitErr *exec.ExitError
+	if errors.As(err, &exitErr) {
+		if stderr := c.stderrExcerpt(exitErr.Stderr); len(stderr) != 0 {
+			cmdErr.msg += ": " + stderr
+		}
+		if exitErr.Exited() {
+			cmdErr.msg += fmt.Sprintf(": exit code %d", exitErr.ExitCode())
+			return cmdErr
+		}
+		if stat, ok := exitErr.Sys().(syscall.WaitStatus); ok && stat.Signaled() {
+			cmdErr.msg += fmt.Sprintf(": exit due to signal %d (%[1]s)", stat.Signal())
+			return cmdErr
+		}
+	}
+
+	if !errors.Is(err, ctx.Err()) {
+		cmdErr.msg += ": " + err.Error()
+	}
+	return cmdErr
+}
+
+func (*cmd) stderrExcerpt(stderr []byte) string {
+	stderr = bytes.TrimSpace(stderr)
+	if len(stderr) == 0 {
+		return ""
+	}
+
+	lines := bytes.Split(stderr, []byte("\n"))
+	slices.Reverse(lines)
+	for _, line := range lines {
+		line = bytes.TrimSpace(line)
+		after, found := bytes.CutPrefix(line, []byte("error: "))
+		if !found {
+			continue
+		}
+		after = bytes.TrimSpace(after)
+		if len(after) == 0 {
+			continue
+		}
+		stderr = after
+		break
+
+	}
+
+	excerpt := string(stderr)
+	if !strconv.CanBackquote(excerpt) {
+		quoted := strconv.Quote(excerpt)
+		excerpt = quoted[1 : len(quoted)-1]
+	}
+	return excerpt
+}
+
+type cmdArgs []any
```

Comment:
Good question and something I should've explained in the description. I'd like to handle types other than strings so they can be logged. For example, if this command fails:

```go
ref, _ := flake.ParseRef("nixpkgs#package")
cmd := nix.Command("build", "--no-link", ref)
cmd.Run(ctx)
```

Then `cmd.Run` knows that the error is related to a nixpkgs package because it has the original `flake.Ref` type. From there, it can do things like automatically log flakes associated with a failed Nix command in a structured way.